apache/zeppelin
# [Zeppelin-1213] Customize editor configuration

### What is this PR for?
Customize editor configuration
- Each user can have different editor setting(For example, editor theme, indentation, font size, editor print margin, etc)
- Each user can have different editor setting for each interpreter. (E.g %spark -> font size, indent, theme)

### What type of PR is it?
Feature

### Todos
- [x] create editor configuration setting page.
- [x] configure theme
- [x] configure support language
- [x] change font size and indent theme 
- [x] editor config serialize / deserialize 
- [x] preview window.
- [x] create and delete, and save.
- [x] add web socket event for code editor setting.
- [x] apply setting for each interpreter tag
- [x] if show line number setting already exists in paragraph, paragraph setting has priority over editor setting

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-1213

### How should this be tested?
1. click to user name menu in navbar and click to Editor Configuration menu.
1. set up editor setting. (and using preview pargraph)
1. click to save button.
1. move to any paragraph.
1. check to apply setting.
1. check to exists conf/code-editor.json in zeppelin directory.

### Screenshots (if appropriate)

<img width="1274" alt="setup" src="https://cloud.githubusercontent.com/assets/10525473/17056055/e56c1236-504b-11e6-803f-4223c813c99e.png">

![2222](https://cloud.githubusercontent.com/assets/10525473/17055992/68b03ad8-504b-11e6-996e-55cfbd72d1b3.gif)

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no
